### PR TITLE
Index shape, size, empty

### PR DESCRIPTION
**Pull request by sfc-gh-samdesai1**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1458118

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

  implement Index.size, Index.shape, Index.empty

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 1]
```diff
@@ -603,12 +602,12 @@ def ndim(self) -> int:
         return 1
 
     @property
-    @index_not_implemented()
-    def size(self) -> None:
+    @is_lazy_check
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:605]: Why have `is_lazy_check` only on this API?

File: tests/integ/modin/index/test_index_methods.py [hunk 2]
```diff
@@ -11,6 +11,8 @@
 from tests.integ.modin.sql_counter import SqlCounter, sql_count_checker
 from tests.integ.modin.utils import assert_index_equal, assert_series_equal
 
+data = {"col1": [1, 2, 3], "col2": [3, 4, 5], "col3": [5, 6, 7]}
```
Comment (sfc-gh-nkrishna) [inline tests/integ/modin/index/test_index_methods.py:14]: I would actually define this as a pytest.fixture instead and pass it in as a parameter to the functions where it is used

File: tests/integ/modin/index/test_index_methods.py [hunk 3]
```diff
@@ -11,6 +11,8 @@
...
+
 NATIVE_INDEX_TEST_DATA = [
```
Comment (sfc-gh-nkrishna) [inline tests/integ/modin/index/test_index_methods.py:16]: Can we add tests with an Index object has has Nan/Null values as well? Nulls are including in size/shape/empty so the code shouldn't change but it's still good to have the test for future APIs as well.

**Review by sfc-gh-nkrishna — COMMENTED** [review]
Can you add implement `Index.size`, `Index.shape`, `Index.empty` to the PR description? @sfc-gh-vbudati If this PR gets merged before the docs one, can you update the docs accordingly?

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 1 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:605]: we don't need this right. We should do the check in `__len__`.

File: tests/integ/modin/index/test_index_methods.py (hunk 2 above)
Comment (sfc-gh-azhan) [reply tests/integ/modin/index/test_index_methods.py:14]: agree, can you define an empty index for all those tests too?

**Review by sfc-gh-azhan — COMMENTED** [review]
Please add change log and also make sure coordinate with the doc update.

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 1 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:605]: sorry, what is this is_lazy_check annotation for ? the api here (shape and size) doesn't sounds like some API can be lazily done, right?

File: tests/integ/modin/index/test_index_methods.py (hunk 3 above)
Comment (sfc-gh-vbudati) [reply tests/integ/modin/index/test_index_methods.py:16]: does your index object also support having list values? Something like pd.Index([[1, 2], [2, 3], [3, 4]]). If yes, add tests for that as well.

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 4]
```diff
@@ -449,12 +449,11 @@ def dtype(self) -> DtypeObj:
         return self.to_pandas().dtype
 
     @property
-    @index_not_implemented()
-    def shape(self) -> None:
+    def shape(self) -> tuple:
```
Comment (sfc-gh-vbudati) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:452]: Add sections for what is returned and examples for each API you implemented here.

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 1 above)
Comment (sfc-gh-samdesai1) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:605]: the is_lazy_check is a check for whether or not we want to use a query for this API or if we want to use the locally cached native pandas object

File: tests/integ/modin/index/test_index_methods.py (hunk 2 above)
Comment (sfc-gh-samdesai1) [reply tests/integ/modin/index/test_index_methods.py:14]: there is an empty index in the native test data

File: CHANGELOG.md [hunk 5]
```diff
@@ -59,7 +59,7 @@
 - Added support for named aggregations in `DataFrame.aggregate` and `Series.aggregate` with `axis=0`.
 - `pd.read_csv` reads using the native pandas CSV parser, then uploads data to snowflake using parquet. This enables most of the parameters supported by `read_csv` including date parsing and numeric conversions. Uploading via parquet is roughly twice as fast as uploading via CSV.
 - Initial work to support an Index directly in Snowpark pandas. Support for Index as a first-class component of Snowpark pandas is coming soon.
-- Added lazy index constructor and support for len, to_pandas() and names. For `df.index`, Snowpark pandas creates a lazy index object. 
+- Added lazy index constructor and support for len, shape, size, empty, to_pandas() and names. For `df.index`, Snowpark pandas creates a lazy index object.
```
Comment (sfc-gh-nkrishna) [inline CHANGELOG.md:62]:
```suggestion
- Added lazy index constructor and support for `len`, `shape`, `size`, `empty`, `to_pandas()` and `names`. For `df.index`, Snowpark pandas creates a lazy index object. 
```

**Review by sfc-gh-vbudati — APPROVED** [review]
Nice work!

**Review by sfc-gh-nkrishna — COMMENTED** [review]
Can you update the documentation @sfc-gh-samdesai1? You need to rebase since Varnika's index docs PR was merged

**Comment by sfc-gh-vbudati** [comment]
I think your merge/rebase messed up

**Comment by sfc-gh-samdesai1** [comment]
> I think your merge/rebase messed up

yep, it should be fixed now though

File: CHANGELOG.md [hunk 6]
```diff
@@ -61,7 +61,7 @@
 - Added support for named aggregations in `DataFrame.aggregate` and `Series.aggregate` with `axis=0`.
 - `pd.read_csv` reads using the native pandas CSV parser, then uploads data to snowflake using parquet. This enables most of the parameters supported by `read_csv` including date parsing and numeric conversions. Uploading via parquet is roughly twice as fast as uploading via CSV.
 - Initial work to support an Index directly in Snowpark pandas. Support for Index as a first-class component of Snowpark pandas is coming soon.
-- Added lazy index constructor and support for len, to_pandas() and names. For `df.index`, Snowpark pandas creates a lazy index object. 
+- Added lazy index constructor and support for `len`, `shape`, `size`, `empty`, `to_pandas()` and `names`. For `df.index`, Snowpark pandas creates a lazy index object.
```
Comment (sfc-gh-nkrishna) [inline CHANGELOG.md:64]: Can you move this to the last line of the section (just best practice)?

**Review by sfc-gh-nkrishna — APPROVED** [review]
LGTM!

File: CHANGELOG.md (hunk 6 above)
Comment (sfc-gh-samdesai1) [reply CHANGELOG.md:64]: It is the second last line, but the last line is also my changes?
